apps/language_server/lib/language_server/experimental/provider/handlers/goto_implementation.ex

```diff
@@ -0,0 +1,24 @@
+defmodule ElixirLS.LanguageServer.Experimental.Provider.Handlers.GotoImplementation do
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Requests.GotoImplementation
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Responses
+  alias ElixirLS.LanguageServer.Experimental.SourceFile
+  alias ElixirLS.LanguageServer.Experimental.SourceFile.Conversions
+  require Logger
+
+  def handle(%GotoImplementation{} = request, _) do
+    source_file = request.source_file
+    pos = request.position
+
+    locations =
+      source_file
+      |> SourceFile.to_string()
+      |> ElixirSense.implementations(pos.line, pos.character + 1)
```

Comment:
One of the design goals of the experimental branch was to eliminate the ad-hoc `pos.character + 1` and `pos.line + 1` math that is present in the current codebase; this request should have had its positions converted to elixir one-based positions already. Is this not working?

Reply:
If we changed the `pos.character + 1` to `pos.character`, the test won't pass:

<img width="1162" alt="image" src="https://user-images.githubusercontent.com/12830256/223895210-6015e960-38a2-4c30-b1a4-8ec43ffe304d.png">

the ad-hoc  `pos.character + 1` and `pos.line + 1` indeed confusing.

Reply:
this might be an artifact of the testing code then, does this work correctly in the actual language server?

Reply:
Let me check first.

Reply:
I'm still a little unclear on what's happening there.
The request from the client has a field called `lsp`, which is the LSP version of the struct. It also has fields directly on the struct. Can you show me what both of those look like for your request?

Reply:
this the original request from the client:

```
[DEBUG][2023-03-10 08:34:25] .../vim/lsp/rpc.lua:284	"rpc.send"	{
  id = 65,
  jsonrpc = "2.0",
  method = "textDocument/implementation",
  params = {
    position = {
      character = 6,
      line = 22
    },
    textDocument = {
      uri = "file:///Users/scottming/Code/recode/lib/recode/task/alias_expansion.ex"
    }
  }
}
```

and I added some logger at line 12 and line 19 to print the request that convert by us and the response by elixir_sense.

```
[ERROR][2023-03-10 08:29:08] ...lsp/handlers.lua:513	'request: %ElixirLS.LanguageServer.Experimental.Protocol.Requests.GotoImplementation{lsp: %ElixirLS.LanguageServer.Experimental.Protocol.Requests.GotoImplementation.LSP{id: 30, jsonrpc: "2.0", method: "textDocument/implementation", text_document: %Protocol.Types.TextDocument.Identifier%{uri: "file:///Users/scottming/Code/recode/lib/recode/task/alias_expansion.ex"}, position: %Protocol.Types.Position%{character: 6, line: 22}}, position: %ElixirLS.LanguageServer.Experimental.SourceFile.Position{line: 23, character: 6}, source_file: %ElixirLS.LanguageServer.Experimental.SourceFile{uri: "file:///Users/scottming/Code/recode/lib/recode/task/alias_expansion.ex", path: "/Users/scottming/Code/recode/lib/recode/task/alias_expansion.ex", version: 0, dirty?: false, document: %ElixirLS.LanguageServer.Experimental.SourceFile.Document{lines: {{:line, "defmodule Recode.Task.AliasExpansion do", "\\n", 1, true}, {:line, "  @moduledoc \\"\\"\\"", "\\n", 2, true}, {:line, "  Multi aliases makes module uses harder to search for in large code bases.", "\\n", 3, true}, {:line, "", "\\n", 4, true}, {:line, "          # preferred", "\\n", 5, true}, {:line, "          alias Module.Foo", "\\n", 6, true}, {:line, "          alias Module.Bar", "\\n", 7, true}, {:line, "", "\\n", 8, true}, {:line, "          # not preferred", "\\n", 9, true}, {:line, "          alias Module.{Foo, Bar}", "\\n", 10, true}, {:line, "", "\\n", 11, true}, {:line, "  This task rewrites the code when `mix recode` runs with `autocorrect: true`.", "\\n", 12, true}, {:line, "  \\"\\"\\"", "\\n", 13, true}, {:line, "", "\\n", 14, true}, {:line, "  use Recode.Task, correct: true, check: true", "\\n", 15, true}, {:line, "", "\\n", 16, true}, {:line, "  alias Recode.Issue", "\\n", 17, true}, {:line, "  alias Recode.Task.AliasExpansion", "\\n", 18, true}, {:line, "  alias Rewrite.Source", "\\n", 19, true}, {:line, "  alias Sourceror.Zipper", "\\n", 20, true}, {:line, "", "\\n", 21, true}, {:line, "  @impl Recode.Task", "\\n", 22, true}, {:line, "  def run(source, opts) do", "\\n", 23, true}, {:line, "    {zipper, issues} =", "\\n", 24, true}, {:line, "      source", "\\n", 25, true}, {:line, "      |> Source.ast()", "\\n", 26, true}, {:line, "      |> Zipper.zip()", "\\n", 27, true}, {:line, "      |> Zipper.traverse([], fn zipper, issues ->", "\\n", 28, true}, {:line, "        expand_alias(zipper, issues, opts[:autocorrect])", "\\n", 29, true}, {:line, "      end)", "\\n", 30, true}, {:line, "", "\\n", 31, true}, {:line, "    case opts[:autocorrect] do", "\\n", 32, true}, {:line, "      true ->", "\\n", 33, true}, {:line, "        Source.update(source, AliasExpansion, ast: Zipper.root(zipper))", "\\n", 34, true}, {:line, "", "\\n", 35, true}, {:line, "      false ->", "\\n", 36, true}, {:line, "        Source.add_issues(source, issues)", "\\n", 37, ...}, {:line, "    end", "\\n", ...}, {:line, "  end", ...}, {:line, ...}, {...}, ...}, starting_index: 1}}, id: 30, jsonrpc: "2.0", method: "textDocument/implementation", text_document: %Protocol.Types.TextDocument.Identifier%{uri: "file:///Users/scottming/Code/recode/lib/recode/task/alias_expansion.ex"}}'
[ERROR][2023-03-10 08:29:08] ...lsp/handlers.lua:513	"elixir_sense_locations: []"
```

In the position field, you can find that:

```elixir
%ElixirLS.LanguageServer.Experimental.SourceFile.Position{line: 23, character: 6}
```

seems like we only converted line not the character, so we need to `+ 1`